models/constraints.py

```diff
@@ -0,0 +1,172 @@
+                     str(shift.end_date_and_time_of_shift.date()), shift.type.value)
+                ] = cp_model.NewBoolVar(
+                        f"shift_employee{employee.id}_shift_start_date{str(shift.start_date_and_time_of_shift.date())}_shift_end_date{str(shift.end_date_and_time_of_shift.date())}_shift{shift.type.value}")
+
+        return all_possible_shift_employee_combinations
+
+    # A constraint that ensures that there will be only one employee in each shift per day
+
+    def add_only_one_employee_in_each_shift_constraint(self):
+        for shift in self.shifts:
+            self.cp_model.AddExactlyOne(self.all_possible_shift_employee_combinations[
+                                            (employee.id,
+                                             str(shift.start_date_and_time_of_shift.date()),
+                                             str(shift.end_date_and_time_of_shift.date()),
+                                             shift.type.value)]
+                                        for employee in self.employees)
+
+    # A constraint that ensures that each employee works at most one shift per day
+    def add_at_most_one_shift_that_starts_in_the_same_day_constraint(self):
+
+        days_in_the_given_shift_list = list(set(str(shift.start_date_and_time_of_shift.date()) for shift in self.shifts))
+
+        for day in days_in_the_given_shift_list:
+            for employee in self.employees:
+                # a list of all the shifts that the employee is working on that day
+                shifts_on_day = []
+
+                for shift in self.shifts:
+                    if str(shift.start_date_and_time_of_shift.date()) == day:
+                        shifts_on_day.append(
+                            self.all_possible_shift_employee_combinations[
+                                (employee.id,
+                                 str(shift.start_date_and_time_of_shift.date()),
+                                 str(shift.end_date_and_time_of_shift.date()),
+                                 shift.type.value)])
+                self.cp_model.AddAtMostOne(shifts_on_day)
+
+    # A constraint that ensures that on each day,
+    def add_constraint_to_prevent_new_employees_working_together(self, first_shift: ShiftTypesEnum, second_shift:ShiftTypesEnum):
+        days_in_the_given_shift_list = list(set(str(shift.start_date_and_time_of_shift.date()) for shift in self.shifts))
+
+        for day in days_in_the_given_shift_list:
+            # collecting all the new employees that are working on this day is specific shifts.
+            new_employees_on_first_shift = []
+            new_employees_on_second_shift = []
+
+            for employee in self.employees:
+                for shift in self.shifts:
+                    if str(shift.start_date_and_time_of_shift.date()) == day and shift.type == first_shift and employee.status == EmployeeStatusEnum.new_employee:
```

Comment:
these two lines are very similar. share more code. you can write nested functions for this with useful names